src/LinearAlgebra/SparsityPattern/VariableImplementation.hpp

```diff
@@ -0,0 +1,214 @@
+/*!
+    \file VariableImplementation.hpp
+*/
+
+#pragma once
+
+namespace GridKit
+{
+namespace Sparse
+{
+    /*!
+        \brief Returns the list of derivatives.
+    */
+    const Variable::DependencyMap& Variable::getDependencies() const
+    {
+        assert(dependencies_ != 0);
+        return *dependencies_;
+    }
+    
+
+    /*!
+        \brief Registers a variable as an unknown of the system and 
+        adds a pointer to the global @a x vector.
+    */
+    void Variable::registerVariable(std::vector<Variable*>& x, 
+                                    const size_t& offset)
+    {
+        setVariableNumber(offset);  // define global variable number
+        setFixed(false);            // not a constant
+
+        x[offset] = this;
+    }
+    
+
+    /*!
+        \brief Adds all dependencies of v to *this.
+    */
+    void Variable::addDependencies(const Variable& v)
+    {
+        for (auto& p : *(v.dependencies_))
+            (*dependencies_)[p.first] = p.second;
+    }
+    
+
+    /*!
+        \brief Multiplies each partial derivative of @a this by @a c. 
+    */
+    void Variable::scaleDependencies(double c)
+    {
+        for (auto& p : *dependencies_)
+            (*dependencies_)[p.first] *= c;
+    }
+    
+    
+    /*!
+        \brief Prints the value and input set of the variable.
+    */
+    void Variable::print(std::ostream& os) const
+    {
+        os << value_;
+        
+        if (isFixed_)
+        {
+            os << " (fixed)";
+            return;
+        }
+        
+        if (variableNumber_ != INVALID_VAR_NUMBER)
+        {
+            os << " (variable " << variableNumber_ << ")";
+            return;
+        }
+        
+        if (dependencies_!= NULL && !dependencies_->empty())
+        {
+            os << " dependencies: [ ";
+            for (auto& p : *dependencies_)
+                os << "(" << p.first << ", " << p.second << ") ";
+            os << "]";
+        }
+    }
+    
+    
+    //------------------------------------
+    // compound assignment operators
+    //------------------------------------
+    
+    /*!
+        \brief Compound addition-assignment operator. Right hand 
+        side is a built-in double type.
+    */
+    Variable& Variable::operator+=(const double& rhs)
+    {
+        value_ += rhs;
+        return *this;
+    }
+    
+    /*!
+        \brief Compound addition-assignment operator. Right hand side 
+        is Variable type.
+    */
+    Variable& Variable::operator+=(const Variable& rhs)
+    {
+        // compute partial derivatives of *this
+        for (auto& p : *(rhs.dependencies_))
+            (*dependencies_)[p.first] += (p.second);
+
+        // compute value of *this
+        value_ += rhs.value_;
+
+        return *this;
+    }
+    
+    // -=
+    /*!
+        \brief Compound substraction-assignment operator. Right hand 
+        side is a built-in double type.
```

Comment:
```suggestion
        \brief Compound subtraction-assignment operator. Right hand 
        side is a built-in double type.
```
"Substraction" is a common error. Perhaps fix with search and replace.